Update external login provider examples with new helper code (#6111)
umbraco/UmbracoDocs: 14/umbraco-cms/reference/security/external-login-providers.md

The process requires adding a couple of new classes (`.cs` files) to your Umbraco project:

* **Custom-named configuration** to add additional configuration for handling different options related to the authentication. [See a generic example of the configuration class to learn more.](#custom-named-configuration)
* A **composer and named** to extend on the default authentication implementation in Umbraco CMS for either Users or Members. [See a generic example to learn more.](#generic-backoffice-login-provider-composer)

> You can setup similar behaviour using a [static extension class](#static-extension-class) and adding them straight into the `Program.cs` file. But you will loose access to dependency injection this way and thus our helper class.

Reviewer: ```suggestion
You can setup similar behavior using a [static extension class](#static-extension-class) and add them straight into the `Program.cs` file. But you will lose access to dependency injection this way, thus our helper class.
```

namespace MyUmbracoProject.CustomAuthentication;

> public class GenericBackOfficeExternalLoginProviderOptions : IConfigureNamedOptions<BackOfficeExternalLoginProviderOptions>

Reviewer: the code title needs to be changed as well to GenericBackOfficeExternalLoginProviderOptions

> public class GenericBackOfficeExternalLoginProviderOptions : IConfigureNamedOptions<BackOfficeExternalLoginProviderOptions>

Reviewer: In this code example am getting 2 errors:
- regarding Icon:

![image](https://github.com/umbraco/UmbracoDocs/assets/83591955/81355d01-e8ea-4a2f-be2a-b98942ac3a2a)

- regarding AutoRedirectLoginToExternalProvider:

![image](https://github.com/umbraco/UmbracoDocs/assets/83591955/039c223e-ae6a-4b31-aa80-f878609f1f00)

Author: The Generic one will not work, its just a code example it has never worked as there is no actual generic provider (I tried the old example just to be sure 😅

Reviewer: these 2 do not exist anymore and can be removed

{% endtabs %}

### Generic backoffice login provider composer
A composer and `genericAuthenticationOptions` configuration class to setup the authentication options for the generic authentication provider using dependency injection.

{% code title="GenericBackOfficeExternalLoginComposer.cs" lineNumbers="true" %}
```csharp
// this example uses a non existing generic OathProvider NuGet package
// but any package that uses the Microsoft.AspNetCore.Authentication.OAuth.OathOptions will work
using AspNet.Security.OAuth.Generic;
using Microsoft.Extensions.Options;
using Umbraco.Cms.Api.Management.Security;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Web.Common.Helpers;

> namespace MyUmbracoProject.CustomAuthentication

Reviewer: ```suggestion
namespace MyUmbracoProject.CustomAuthentication;
```

public class GenericBackOfficeExternalLoginComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        // register the generic loginProvider options (auto linking, manual linking, ...)
        builder.Services.ConfigureOptions<GenericBackOfficeExternalLoginProviderOptions>();
        // register the generic authtication options (secret, callback, errorhandling, ...)
        builder.Services.ConfigureOptions<ConfigureGenericAuthenticationOptions>();

        builder.AddBackOfficeExternalLogins(logins =>
        {
            logins.AddBackOfficeLogin(
                backOfficeAuthenticationBuilder =>
                {

>                     // this Add... method will be part of the OathProvider nuget package you install

Reviewer: add own provider method

>                     backOfficeAuthenticationBuilder.AddGenericProvider(

Reviewer: am getting the following error on AddGenericProvider here:

![image](https://github.com/umbraco/UmbracoDocs/assets/83591955/a13f5ebf-61bb-473b-8e11-5667407ef726)

>                         BackOfficeAuthenticationBuilder.SchemeForBackOffice(GenericOfficeExternalLoginProviderOptions

Reviewer: ```suggestion
                        BackOfficeAuthenticationBuilder.SchemeForBackOffice(GenericBackOfficeExternalLoginProviderOptions
```

>                         BackOfficeAuthenticationBuilder.SchemeForBackOffice(GenericOfficeExternalLoginProviderOptions

Reviewer: ```suggestion
                        BackOfficeAuthenticationBuilder.SchemeForBackOffice(GenericBackOfficeExternalLoginProviderOptions
```

                            .SchemeName)!,
                        options =>
                        {
                            // need to give an empty action here for the options pattern configuration to work 🤷
                            // if you do not wish to use the umbraco default error handling and hardcode all your values instead of injecting them,
                            // you can set the configuration right here instead.
                        });
                });
        });
    }
}

// the ...AuthenticationOptions method will be part of the OathProvider nuget package you install
// check the Add... method invoked on the backOfficeAuthenticationBuilder to figure out the correct type

> public class ConfigureGenericAuthenticationOptions : IConfigureNamedOptions<GenericAuthenticationOptions>

Reviewer: am getting the following error for GenericAuthenticationOptions, is the same on your side?

![image](https://github.com/umbraco/UmbracoDocs/assets/83591955/8bf845dd-148b-4912-9cd3-ece7e2848fa8)

{% endtabs %}

### Generic backoffice login provider composer

> A composer and genericAuthenticationOptions configuration class to setup the authentication options for the generic authentication provider using dependency injection

Reviewer: ```suggestion
A composer and `genericAuthenticationOptions` configuration class to setup the authentication options for the generic authentication provider using dependency injection.
```

{% code title="GenericBackOfficeExternalLoginComposer.cs" lineNumbers="true" %}
```csharp

> // this example uses a non existing generic OathProvider nuget package

Reviewer: ```suggestion
// this example uses a non existing generic OathProvider NuGet package
```

                {
                    // The scheme must be set with this method to work for the back office
                    var schemeName =

>                         backOfficeAuthenticationBuilder.SchemeForBackOffice(GenericOfficeExternalLoginProviderOptions

Reviewer: getting the following error on GenericOfficeExternalLoginProviderOptions, is it the same for you?

![image](https://github.com/umbraco/UmbracoDocs/assets/83591955/b926f509-f8c4-4963-9380-f49b7a0fbac3)

                            .SchemeName);

                    ArgumentNullException.ThrowIfNull(schemeName);

>                     backOfficeAuthenticationBuilder.AddGenericProvider(

Reviewer: getting the following error on AddGenericProvider, is it the same for you?

![image](https://github.com/umbraco/UmbracoDocs/assets/83591955/be482fb3-6688-4e18-849f-c96e7c6dcf5a)
